### ISSUE TITLE
Optimizer other than GradientDescent throwing errors on first run

### ISSUE THREAD

**Issue by dsgupta** [tensorflow/tensorflow#31426]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colag
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below): 1.14
- Python version: 3.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the current behavior**
On running the train method, it is throwing error despite initializing all variables before the optimize step. 

**Describe the expected behavior**
It should not throw errors and initialize all variables.
**Code to reproduce the issue**
Colab [gist](https://colab.research.google.com/gist/dsgupta/bb90037cb95573f610ff2d974883286d/test.ipynb)

**Comment by ymodak** [comment]
You need to initialize the variables inside the session. For instance,
```python
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
```
See https://www.tensorflow.org/guide/variables#initializing_variables
https://colab.research.google.com/gist/dsgupta/bb90037cb95573f610ff2d974883286d/test.ipynb#scrollTo=du7QsOQeKpAS&line=8&uniqifier=1

**Comment by dsgupta** [comment]
@ymodak  
Doesn't the following take care of that?
```
with tf.Session() as sess:
    sess.run([var_init, iterator_init])
```

**Comment by ymodak** [comment]
You need to slightly tweak the ```fetches``` arg in ```session.run```
The ```fetches``` argument may be a single graph element, or an arbitrarily nested list, tuple.
See https://www.tensorflow.org/api_docs/python/tf/Session#run
You need to pass elements ```var_init, iterator_init``` as a tuple perhaps.
```python
my_tuple = (var_init, iterator_init)
with tf.Session(my_tuple) as sess:
    sess.run()
```

**Comment by isaprykin** [comment]
I think it's because global_variables() are read by global_variables_initializer before they are created by the train_step.

This works:
with mirrored_strategy.scope():
  input_iterator = dist_dataset.make_initializable_iterator()
  iterator_init = input_iterator.initialize()
  loss = train_step(input_iterator.get_next())
  var_init = tf.global_variables_initializer()  
  with tf.Session() as sess:
    sess.run([var_init, iterator_init])
    for _ in range(10):
      print(sess.run(loss))


Thanks for the colab.
